contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,456 @@
+        uint256 wethAmount;
+
+        // create CDP, lock PETH, and draw DAI
+        sender.cdpID = tub.open();
+        tub.lock(sender.cdpID, pethAmount);
+        tub.draw(sender.cdpID, daiAmount);
+
+        //trade DAI for PETH and reinvest into CDP for # of layers
+        for (uint256 i = 0; i < sender.layers - 1; i++) {
+            wethAmount = marketBuy(weth, dai, daiAmount);
+            pethAmount = weth2peth(wethAmount);
+
+            tub.join(pethAmount);
+            tub.lock(sender.cdpID, pethAmount);
+
+            daiAmount = wdiv(wmul(currPricePeth, pethAmount), collatRatio);
+            tub.draw(sender.cdpID, daiAmount);
+        }
+
+        sender.daiAmountFinal = daiAmount;
+        sender.totalDebt = wdiv(tub.tab(sender.cdpID),ray2wad(tub.chi()));
+
+        investors[msg.sender] = sender;
+
+        emit OpenPosition(msg.sender, msg.value, sender.purchPrice, _layers);
+    }
+
+
+    /**
+    * @dev Liquidate - Liqduiates the sender's position and returns ETH
+    * Possible outcomes:
+    * 1. USD/ETH drops below sender's price floor
+    * 2. USD/ETH appreciates relative to sender's purchase price
+    * 3. USD/ETH deppreciates relative to sender's purchase price
+    * Sender must send ETH with function call to cover the governance and stability fee with use of the Maker platform
+    * Any excess ETH will be returned with the closing of the position
+    */
+    function liquidate()
+        payable
+        onlyInvestor
+        public
+    {
+        uint256 currPriceEth = uint256(pip.read());                          // DAI/WETH In WAD
+
+        Investor memory sender = investors[msg.sender];
+
+        uint256 releasedPeth;
+        uint256 remainingPeth = tub.ink(sender.cdpID);                       // retrieve value of collatoral in CDP
+        uint256 payout;
+        uint256 wethAmount;
+        uint256 daiAmount = sender.daiAmountFinal;
+
+        // checks if CDP debt is 0
+        if (tub.tab(sender.cdpID) == 0) {
+            //*** CDP is auto-liquidated
+            //*** Convert to WETH and send back to investor = 1 + 2
+            //*** 1. Remaining PETH in debt-free CDP
+            //*** 2. Outstanding DAI from final layer
+
+            releaseWeth(sender.cdpID, remainingPeth);
+            wethAmount = marketBuy(weth, dai, sender.daiAmountFinal);
+            payout = add(remainingPeth,wethAmount);
+
+        } else {
+
+            //*** USD/ETH price deppreciated OR appreciated
+
+            require(0 < msg.value, "Ether is required to liqudiate.");
+
+            uint256 remainingDebt = sender.totalDebt;
+            uint256 remainingDai;
+            uint256 excessWeth;
+
+            IWETH(weth).deposit.value(msg.value)();
+
+            uint256 mkrFee;
+            uint256 wethFee;
+            (mkrFee, wethFee) = wipeFee(sender.cdpID, remainingDebt);
+
+            // Pay for governance fee
+            require(wethFee <= msg.value, "Not enough ether provided for fees");
```

Comment:
How's the user know how much ether to send back for a successful liquidation? Is this something that should be added to the frontend?

Reply:
It's something that will be calculated off-chain and provided to the user through the front-end. If the user sends more than what is required for the fee, the excess WETH will be sent back within the `payout` calculation